### PR TITLE
Deprecate early definitions under -Xsource:2.14

### PR DESCRIPTION
**Pull request by som-snytt**
Under -Xsource:2.14, offer trait parameters,
otherwise the mirage of a distant oasis.

Eerily, github nailed the suggested reviewer again. (In the sense of "hit the nail on the head.")

Cf https://github.com/scala/scala/pull/6349

Fixes scala/scala-dev#513

### CODE REVIEW

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala
Comment (eed3si9n) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: Given the discussion in https://github.com/scala/scala-dev/issues/513, it might be good to tell people that you don't have to wait till trait parameters land to start the migration.

How about:

```
t2796.scala:9: warning: early initializers are deprecated; avoid using val or var in traits.
trait T1 extends {
                 ^
```

Comment (smarter) [reply src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: I think the most user-friendly thing to do would be to add a documentation page that helps users transition away from early initializers, and link to that page from the error message.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: `Don't wait! Migrate now! 1-800-MIGRATE!`

**Review by adriaanm — APPROVED** [review]
Baby steps do tend to go in all directions, as does our build with deprecation warning cleanliness. But! But! My global, where do I initialize my single global?

**Comment by adriaanm** [comment]
We usually don't deprecate without alternatives, but this has a few acceptable ones, and I think their overall usage is low.

Here's another neat trick to avoid them (🎩-tip @gsps):
https://github.com/epfl-lara/inox/pull/71/commits/d13789665dbc9b8ce7a32557d68460d11307fa67:
`new T { val early: Early = e } with U` becomes `class TParam(val early: Early) extends U; new TParam(e))`

**Comment by som-snytt** [comment]
Hope to contribute the noise-cancelling reporter soon, if the ergonomics work out.

**Comment by xeno-by** [comment]
@som-snytt @adriaanm Let's put this deprecation under `-Xsource:2.14`? For those with `-Xfatal-warnings` enabled, this is a breaking language change, and I think that changes like that need an approval from the SIP committee.

**Comment by som-snytt** [comment]
I wonder if `-Xmigration` is applicable? That would be a way to ask for a heads up without committing to what used to be `-Xfuture`. Normally, `-Xsource:2.14` would mean `unsupported feature`.

**Comment by adriaanm** [comment]
I don’t think this needs to weigh on the SIP meeting agenda, which is
already super packed. We already approved, IIRC, that early definitions are
not in 3.0. Thus we need a deprecation warning. I would rather announce the
breakage early and loudly than silently and without recourse at the last
minute.

For those big shops that rely crucially on xfatalwarnings, it would be
pretty useful to invest in warning filtering. Everything is in place for
configurable error reporting.

**Comment by nafg** [comment]
I'm sure it's not just big shops that use -Xfatal-warnings, but I guess the other type that use it probably either would never use early initializers or would be happy to remove them?

**Comment by adriaanm** [comment]
Well, a small shop that really is bothered by how coarse-grained -Xfatal-warnings is, could also help out with a PR that improves our warning filterting :-)

**Comment by SethTisue** [comment]
this looks pretty nice for warning suppression: https://github.com/ghik/silencer/pull/14

**Pull request merged**
